File: hard/25. Reverse Nodes in k-Group/25. Reverse Nodes in k-Group.ts
/**
 * Definition for singly-linked list.
 */

class ListNode {
    val: number
    next: ListNode | null
    constructor(val?: number, next?: ListNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.next = (next === undefined ? null : next)
    }
}

const reverse = (start: ListNode | null, k: number): [(ListNode | null), (ListNode | null)] => {
    let prev: ListNode | null = null, curr = start
    while (k && curr) {
        const nxt: ListNode | null = curr.next
        curr.next = prev
        prev = curr
        curr = nxt
        k -= 1
    }

    return [prev, curr]
}

function reverseKGroup(head: ListNode | null, k: number): ListNode | null {
    let length = 0
    let curr = head
    while (curr) {
        length += 1
        curr = curr.next
    }

    const dummy = new ListNode(0)
    dummy.next = head
    let groupTail: ListNode | null = dummy
    curr = head

    while (length >= k) {
        const tail = curr
        const [reversedHead, nextGroupTail] = reverse(curr, k)

        if (groupTail)
            groupTail.next = reversedHead

        if (tail)
            tail.next = nextGroupTail

        groupTail = tail
        curr = nextGroupTail
        length -= k
    }

    return dummy.next
};

export { }

File: hard/25. Reverse Nodes in k-Group/25. Reverse Nodes in k-Group.py
from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        list_len = 0
        curr = head
        while curr:
            curr = curr.next
            list_len += 1

        curr = head

        new_list = ListNode(0)
        curr_new = new_list

        for _ in range(0, (list_len // k) * k, k):
            if not curr or not curr.next:
                break

            next_val: Optional[ListNode] = None

            step = 0
            while step < k and curr:
                next_val = ListNode(curr.val, next_val)
                curr = curr.next
                step += 1

            curr_new.next = next_val

            while curr_new.next:
                curr_new = curr_new.next

        if curr:
            curr_new.next = curr

        return new_list.next


solution = Solution()

ex1_5 = ListNode(5)
ex1_4 = ListNode(4, ex1_5)
ex1_3 = ListNode(3, ex1_4)
ex1_2 = ListNode(2, ex1_3)
ex1_1 = ListNode(1, ex1_2)

example_1 = ex1_1

res = solution.reverseKGroup(example_1, 2)

while res:
    print(res.val)
    res = res.next

# print(solution.reverseKGroup(example_2, 3))
